Implement AlpakaBackendProducer and AlpakaBackendFilter [14.0.x]

fwyzard opened the pull request:
#### PR description:

Implement `AlpakaBackendProducer`, an empty alpaka-based `EDProducer` whose only purpose is to save in the event what alpaka backend has been used.

Implement `AlpakaBackendFilter`, an `EDFilter` that selects events based on the alpaka backend used to run a previous producer.

Implement a unit test for both modules.

The aim is to replace this `SwitchProducer` in the HLT menu
```python

process.statusOnGPU = SwitchProducerCUDA(
   cpu = cms.EDProducer( "BooleanProducer",
       value = cms.bool( False )
   ),
  cuda = cms.EDProducer( "BooleanProducer",
       value = cms.bool( True )
   ),
 )

process.statusOnGPUFilter = cms.EDFilter( "BooleanFilter",
    src = cms.InputTag( "statusOnGPU" )
)

process.Status_OnCPU = cms.Path( process.statusOnGPU + ~process.statusOnGPUFilter )
process.Status_OnGPU = cms.Path( process.statusOnGPU + process.statusOnGPUFilter )
```
with this alpaka-based solution
```python
process.hltBackend = cms.EDProducer('AlpakaBackendProducer@alpaka')

process.hltStatusOnGPUFilter = cms.EDFilter('AlpakaBackendFilter',
  producer = cms.InputTag('hltBackend', 'backend'),
  backends = cms.vstring('CudaAsync', 'ROCmAsync')
)

process.Status_OnCPU = cms.Path( process.hltBackend + ~process.hltStatusOnGPUFilter )
process.Status_OnGPU = cms.Path( process.hltBackend + process.hltStatusOnGPUFilter )
```

#### PR validation:

The new unit test passes.

#### Backport status

Backport of #44387 to 14.0.x for data taking.

fwyzard committed:
Implement AlpakaBackendProducer and AlpakaBackendFilter

Implement AlpakaBackendProducer, an empty alpaka-based EDProducer whose
only purpose is to save in the event what alpaka backend has been used.

Implement AlpakaBackendFilter, an EDFilter that selects events based on
the alpaka backend used to run a previous producer.

Implement a unit test for both modules.

fwyzard commented:
backport #44387

fwyzard commented:
enable gpu

fwyzard commented:
please test

cmsbuild commented:
A new Pull Request was created by  @fwyzard for CMSSW_14_0_X.

It involves the following packages:

- HeterogeneousCore/AlpakaCore (**heterogeneous**)


@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@missirol, @rovere, @makortel this is something you requested to watch as well.
@rappoccio, @sextonkennedy, @antoniovilela you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"8d6395cbea0988d37dbe4c5f45c391a6a030123b":{"files":["HeterogeneousCore/AlpakaCore/plugins/AlpakaBackendFilter.cc","HeterogeneousCore/AlpakaCore/plugins/BuildFile.xml","HeterogeneousCore/AlpakaCore/plugins/alpaka/AlpakaBackendProducer.cc","HeterogeneousCore/AlpakaCore/test/BuildFile.xml","HeterogeneousCore/AlpakaCore/test/testAlpakaBackendFilter.py"],"squashed":false,"time":1710322854}},"emoji":{"1994102312":"+1","1994102474":"+1"},"last_seen_sha":"8d6395cbea0988d37dbe4c5f45c391a6a030123b","signatures":{}} -->

fwyzard commented:
@mmusich @Martin-Grunewald @missirol FYI

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a818d9/38096/summary.html
**COMMIT**: 8d6395cbea0988d37dbe4c5f45c391a6a030123b
**CMSSW**: CMSSW_14_0_X_2024-03-12-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44388/38096/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-12-2300+a818d9/61644/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 64 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-12-2300+a818d9/61644/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3342271
 - DQMHistoTests: [Total failures: 1358](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-12-2300+a818d9/61644/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3340891
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-12-2300+a818d9/61644/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 24 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-12-2300+a818d9/61640/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 584](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-12-2300+a818d9/61640/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39156
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-12-2300+a818d9/61640/triggerResults): no differences found

makortel reviewed (commented):
The test doesn't seem to really test the behavior of `AlpakaBackendFilter`.

fwyzard commented:
> The test doesn't seem to really test the behavior of `AlpakaBackendFilter`.

What do you mean ?

makortel commented on HeterogeneousCore/AlpakaCore/test/testAlpakaBackendFilter.py:
```diff
@@ -0,0 +1,41 @@
+  backends = cms.vstring(backend)
+)
+
+process.SelectedBackend = cms.Path(process.alpakaBackendProducer + process.alpakaBackendFilter)
+process.AnyOtherBackend = cms.Path(process.alpakaBackendProducer + ~process.alpakaBackendFilter)
```
> > The test doesn't seem to really test the behavior of `AlpakaBackendFilter`.
> 
> What do you mean ?

I mean the filtering decision being the expected one is not being tested. E.g. something along
```suggestion
process.mustRun = cms.EDProducer("edmtest::MustRunIntProducer", ivalue=cms.int32(1))
process.mustNotRun = cms.EDProducer("FailingProducer")

process.SelectedBackend = cms.Path(process.alpakaBackendProducer + process.alpakaBackendFilter + process.mustRun)
process.AnyOtherBackend = cms.Path(process.alpakaBackendProducer + ~process.alpakaBackendFilter + process.mustNotRun)
```
would lead to the job to fail if the filtering decision (i.e. the joint logic in `AlpakaBackendFilter` constructor and `filter()` function) would be unexpected for any reason.

fwyzard replied:
Ah, I see.
I guess it's enough to have the `FailingProducer` ?
Is that a real module one can use, or just an example ?

makortel replied:
> I guess it's enough to have the `FailingProducer` ?

Right, in this case where the "other behavior" is achieved with `~` (whose behavior we should be testing in the framework) the first case should indeed be sufficient.

If the other Path would use another instance of `AlpakaBackendFilter` with different `backend` configuration, then testing it explicitly could be useful.

> Is that a real module one can use, or just an example ?

It is available (from test modules defined in `FWCore/Framework/test`).

fwyzard replied:
OK, thanks, I'll make the changes.

cmsbuild commented:
Pull request #44388 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_14_0_X IBs after it passes the integration tests and once validation in the development release cycle CMSSW_14_1_X is complete. This pull request will now be reviewed by the release team before it's merged. @rappoccio, @sextonkennedy, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a818d9/38178/summary.html
**COMMIT**: 6c9a2840a10514b10caa5b51fe35f0ea2def4231
**CMSSW**: CMSSW_14_0_X_2024-03-15-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44388/38178/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-15-1100+a818d9/61702/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 70 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-15-1100+a818d9/61702/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3346226
 - DQMHistoTests: [Total failures: 1361](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-15-1100+a818d9/61702/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3344843
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-15-1100+a818d9/61702/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - You potentially [removed 8 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-15-1100+a818d9/61700/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 48 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-15-1100+a818d9/61700/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 1835](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-15-1100+a818d9/61700/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 37905
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-15-1100+a818d9/61700/triggerResults): no differences found

smuzaffar commented:
@fwyzard , I need to backport build rule change to 14.0.X too

fwyzard commented:
OK, sure.

I'm not pushing for it, since we are testing the same functionality in 14.1.x - but thanks if you decide to backport it.

makortel commented on HeterogeneousCore/AlpakaCore/plugins/BuildFile.xml:
```diff
@@ -3,5 +3,17 @@
+  <flags EDM_PLUGIN="1"/>
+</library>
+
+<library file="alpaka/*.cc" name="HeterogeneousCoreAlpakaTestPluginsPortable">
```
```suggestion
<library file="alpaka/*.cc" name="HeterogeneousCoreAlpakaCorePluginsPortable">
```

fwyzard committed:
Fix plugin library name AlpakaCore/plugins/BuildFile.xml

This name is unique, the previous name clashed with AlpakaTest/plugins/BuildFile.xml

Co-authored-by: Matti Kortelainen <matti.kortelainen@cern.ch>

cmsbuild commented:
Pull request #44388 was updated. @fwyzard, @cmsbuild, @makortel can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a818d9/38208/summary.html
**COMMIT**: c58cda6d29a17193202f859576cdbac3bc379ca2
**CMSSW**: CMSSW_14_0_X_2024-03-17-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/44388/38208/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 19 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-17-2300+a818d9/61743/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 48 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-17-2300+a818d9/61743/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3346212
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-17-2300+a818d9/61743/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3346184
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 205 log files, 166 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-03-17-2300+a818d9/61743/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-17-2300+a818d9/61755/validateJR.html)
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: [Total failures: 23](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-17-2300+a818d9/61755/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39717
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_14_0_X_2024-03-17-2300+a818d9/61755/triggerResults): no differences found

cmsbuild commented:
**REMINDER** @sextonkennedy, @antoniovilela, @rappoccio: This PR was tested with cms-sw/cmssw#44440, please check if they should be merged together

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_14_0_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_14_1_X is complete. This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
